File: src/patterns.ts
/** 
 * Illegal character pattern 
 */
export const ILLEGAL_CHAR = /[^ -~\s]+/;

/** 
 * Rainlang word pattern 
 */
export const WORD_PATTERN = /^[a-z][0-9a-z-]*$/;

/** 
 * Import hash pattern 
 */
export const HASH_PATTERN = /^0x[0-9a-fA-F]{64}$/;

/** 
 * Rainlang numeric pattern 
 */
export const NUMERIC_PATTERN = /^0x[0-9a-fA-F]+$|^0b[0-1]+$|^\d+$|^[1-9]\d*e\d+$/;

/** 
 * Hex pattern 
 */
export const HEX_PATTERN = /^0x[0-9a-fA-F]+$/;

/** 
 * Binary pattern 
 */
export const BINARY_PATTERN = /^0b[0-1]+$/;

/** 
 * e numberic pattern 
 */
export const E_PATTERN = /^[1-9]\d*e\d+$/;

/** 
 * Integer pattern 
 */
export const INT_PATTERN = /^\d+$/;

/** 
 * RainDocument Namespace pattern 
 */
export const NAMESPACE_PATTERN = /^(\.?[a-z][0-9a-z-]*)*\.?$/;

/** 
 * Comment pattern 
 */
export const COMMENT_PATTERN = /\/\*[\s\S]*?(?:\*\/|$)/;

/** 
 * whitespace pattern 
 */
export const WS_PATTERN = /\s+/;

/** 
 * binding dependency pattern 
 */
export const DEP_PATTERN = /'\.?[a-z][0-9a-z-]*(\.[a-z][0-9a-z-]*)*/;

/** 
 * import pattern 
 */
export const IMPORTS_PATTERN = /@/;

/** 
 * binding pattern 
 */
export const BINDING_PATTERN = /#/;

/** 
 * non-empty text pattern 
 */
export const NON_EMPTY_PATTERN = /[^\s]/;

/** 
 * operand arguments pattern 
 */
export const OPERAND_ARG_PATTERN = /^[0-9]+$|^0x[a-fA-F0-9]+$|^'\.?[a-z][a-z0-9-]*(\.[a-z][a-z0-9-]*)*$/;

/** 
 * namespace segments delimitier pattern 
 */
export const NAMESPACE_SEGMENT_PATTERN = /\./;

/** 
 * rainlang source delimiter pattern 
 */
export const SOURCE_PATTERN = /;/;

/** 
 * rainlang line delimiter pattern 
 */
export const SUB_SOURCE_PATTERN = /,/;

/** 
 * any not whitespace pattern 
 */
export const ANY_PATTERN = /\S+/;

/** 
 * rainlang lhs pattern 
 */
export const LHS_PATTERN = /^[a-z][a-z0-9-]*$|^_$/;

/** 
 * Lint patterns for RainDocument/RainlangDocument 
 */
export namespace LintPatterns {
    /** 
     * ignore next line 
     */
    export const IGNORE_NEXT_LINE = /\bignore-next-line\b/;

    /** 
     * ignore words 
     */
    export const IGNORE_WORDS = /\bignore-words\b/;

    /** 
     * ignore undefined words 
     */
    export const IGNORE_UNDEFINED_WORDS = /\bignore-undefined-words\b/;
}

File: src/types/patterns.rs
//! All Regular Expression patterns used for parsing texts

use regex::Regex;
use once_cell::sync::Lazy;

/// Illegal character pattern
pub static ILLEGAL_CHAR: Lazy<Regex> = Lazy::new(|| Regex::new(r"[^ -~\s]+").unwrap());

/// Rainlang word pattern
pub static WORD_PATTERN: Lazy<Regex> = Lazy::new(|| Regex::new(r"^[a-z][0-9a-z-]*$").unwrap());

/// Import hash pattern
pub static HASH_PATTERN: Lazy<Regex> = Lazy::new(|| Regex::new(r"^0x[0-9a-fA-F]{64}$").unwrap());

/// Rainlang numeric pattern
pub static NUMERIC_PATTERN: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"^0x[0-9a-fA-F]+$|^0b[0-1]+$|^\d+$|^[1-9]\d*e\d+$").unwrap());

/// Hex pattern
pub static HEX_PATTERN: Lazy<Regex> = Lazy::new(|| Regex::new(r"^0x[0-9a-fA-F]+$").unwrap());

/// Binary pattern
pub static BINARY_PATTERN: Lazy<Regex> = Lazy::new(|| Regex::new(r"^0b[0-1]+$").unwrap());

/// e numberic pattern
pub static E_PATTERN: Lazy<Regex> = Lazy::new(|| Regex::new(r"^[1-9]\d*e\d+$").unwrap());

/// Integer pattern
pub static INT_PATTERN: Lazy<Regex> = Lazy::new(|| Regex::new(r"^\d+$").unwrap());

/// RainDocument Namespace pattern
pub static NAMESPACE_PATTERN: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"^(\.?[a-z][0-9a-z-]*)*\.?$").unwrap());

/// Comment pattern
pub static COMMENT_PATTERN: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"\/\*[\s\S]*?(?:\*\/|$)").unwrap());

/// whitespace pattern
pub static WS_PATTERN: Lazy<Regex> = Lazy::new(|| Regex::new(r"\s+").unwrap());

/// binding dependency pattern
pub static DEP_PATTERN: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"'\.?[a-z][0-9a-z-]*(\.[a-z][0-9a-z-]*)*").unwrap());

/// import pattern
pub static IMPORTS_PATTERN: Lazy<Regex> = Lazy::new(|| Regex::new("@").unwrap());

/// binding pattern
pub static BINDING_PATTERN: Lazy<Regex> = Lazy::new(|| Regex::new("#").unwrap());

/// non-empty text pattern
pub static NON_EMPTY_PATTERN: Lazy<Regex> = Lazy::new(|| Regex::new(r"[^\s]").unwrap());

/// operand arguments pattern
pub static OPERAND_ARG_PATTERN: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"^[0-9]+$|^0x[a-fA-F0-9]+$|^'\.?[a-z][a-z0-9-]*(\.[a-z][a-z0-9-]*)*$").unwrap()
});

/// namespace segments delimitier pattern
pub static NAMESPACE_SEGMENT_PATTERN: Lazy<Regex> = Lazy::new(|| Regex::new(r"\.").unwrap());

/// rainlang source delimiter pattern
pub static SOURCE_PATTERN: Lazy<Regex> = Lazy::new(|| Regex::new(";").unwrap());

/// rainlang line delimiter pattern
pub static SUB_SOURCE_PATTERN: Lazy<Regex> = Lazy::new(|| Regex::new(",").unwrap());

/// any not whitespace pattern
pub static ANY_PATTERN: Lazy<Regex> = Lazy::new(|| Regex::new(r"\S+").unwrap());

/// rainlang lhs pattern
pub static LHS_PATTERN: Lazy<Regex> = Lazy::new(|| Regex::new(r"^[a-z][a-z0-9-]*$|^_$").unwrap());

/// the default elided binding msg
pub static DEFAULT_ELISION: &str = "elided binding, requires rebinding";

/// Lint patterns for RainDocument/RainlangDocument
pub mod lint_patterns {
    use regex::Regex;
    use once_cell::sync::Lazy;

    /// ignore next line
    pub static IGNORE_NEXT_LINE: Lazy<Regex> =
        Lazy::new(|| Regex::new(r"\bignore-next-line\b").unwrap());

    /// ignore words
    pub static IGNORE_WORDS: Lazy<Regex> = Lazy::new(|| Regex::new(r"\bignore-words\b").unwrap());

    /// ignore undefined words
    pub static IGNORE_UNDEFINED_WORDS: Lazy<Regex> =
        Lazy::new(|| Regex::new(r"\bignore-undefined-words\b").unwrap());
}
